activerecord/lib/active_record/relation/query_methods.rb

```diff
@@ -1829,6 +1839,35 @@ def check_if_method_has_arguments!(method_name, args, message = nil)
         end
       end
 
+      def process_select_args(fields)
+        fields.flat_map do |field|
+          if field.is_a?(Hash)
+            transform_select_hash_values(field)
+          else
+            field
+          end
+        end
+      end
+
+      def transform_select_hash_values(fields)
+        fields.flat_map do |key, columns_aliases|
+          case columns_aliases
+          when Hash
+            columns_aliases.map do |column, column_alias|
+              arel_column("#{key}.#{column}", &:itself).as(column_alias.to_s)
```

Comment:
this `&:itself` is incorrect as it causes `arel_column` to return a String, on which you call `#as`.

I was able to make this fail with this test:

```ruby
    def test_select_with_invalid_nested_field
      assert_raises(ActiveRecord::StatementInvalid) do
        Post.select(posts: { "UPPER(title)" => :post_title }).take
      end
    end
```